waku/waku_rln_relay/group_manager/on_chain/group_manager.nim

```diff
@@ -226,310 +248,185 @@ method withdrawBatch*(
-
-  try:
-    # we always want to sync from last processed block => latest
-    # chunk events
-    while true:
-      # if the fromBlock is less than 2k blocks behind the current block
-      # then fetch the new toBlock
-      if fromBlock >= currentLatestBlock:
-        break
-
-      if fromBlock + blockChunkSize > currentLatestBlock:
-        g.retryWrapper(currentLatestBlock, "Failed to get the latest block number"):
-          cast[BlockNumber](await ethRpc.provider.eth_blockNumber())
-
-      let toBlock = min(fromBlock + blockChunkSize, currentLatestBlock)
-      debug "fetching events", fromBlock = fromBlock, toBlock = toBlock
-      await sleepAsync(rpcDelay)
-      futs.add(g.getAndHandleEvents(fromBlock, toBlock))
-      if futs.len >= maxFutures or toBlock == currentLatestBlock:
-        await g.batchAwaitBlockHandlingFuture(futs)
-        g.setMetadata(lastProcessedBlock = some(toBlock)).isOkOr:
-          error "failed to persist rln metadata", error = $error
-        futs = newSeq[Future[bool]]()
-      fromBlock = toBlock + 1
-  except CatchableError:
-    raise newException(
-      CatchableError,
-      "failed to get the history/reconcile missed blocks: " & getCurrentExceptionMsg(),
-    )
-
-  # listen to blockheaders and contract events
-  try:
-    await g.startListeningToEvents()
-  except CatchableError:
-    raise newException(
-      ValueError, "failed to start listening to events: " & getCurrentExceptionMsg()
-    )
+  # Set up the polling interval - more frequent to catch roots
+  const rpcDelay = 1.seconds
 
-method startGroupSync*(
-    g: OnchainGroupManager
-): Future[GroupManagerResult[void]] {.async.} =
-  ?resultifiedInitGuard(g)
-  # Get archive history
-  try:
-    await startOnchainSync(g)
-    return ok()
-  except CatchableError, Exception:
-    return err("failed to start group sync: " & getCurrentExceptionMsg())
-
-method onRegister*(g: OnchainGroupManager, cb: OnRegisterCallback) {.gcsafe.} =
-  g.registerCb = some(cb)
+  info "Starting to track Merkle root changes"
 
-method onWithdraw*(g: OnchainGroupManager, cb: OnWithdrawCallback) {.gcsafe.} =
-  g.withdrawCb = some(cb)
+  while true:
+    try:
+      # Fetch the current root
+      let rootRes = await g.fetchMerkleRoot()
+      if rootRes.isErr():
+        error "Failed to fetch Merkle root", error = rootRes.error
+        await sleepAsync(rpcDelay)
+        continue
+
+      let currentRoot = toMerkleNode(rootRes.get())
+
+      if g.validRoots.len == 0 or g.validRoots[g.validRoots.len - 1] != currentRoot:
+        let overflowCount = g.validRoots.len - AcceptableRootWindowSize + 1
+        if overflowCount > 0:
+          for i in 0 ..< overflowCount:
+            discard g.validRoots.popFirst()
+
+        g.validRoots.addLast(currentRoot)
+        info "Detected new Merkle root",
+          root = currentRoot.toHex, totalRoots = g.validRoots.len
+
+      let proofResult = await g.fetchMerkleProofElements()
```

Comment:
If the `root` didn't change, there's no need to fetch a new merkle proof - it would be the same as before.